### ISSUE TITLE
Add Unit Test Manager

### ISSUE THREAD

**Issue by nahum-test** [mercere99/Empirical#1]

**Comment by nahumj** [comment]
Proposal for Catch
https://github.com/philsquared/Catch

**Comment by bocajnotnef** [comment]
+1 to Catch, will implement

**Comment by bocajnotnef** [comment]
Additionally, need a way to metric test coverage.

**Comment by bocajnotnef** [comment]
So, initial thoughts: we could use the gcov utility to metric code coverage of the plain c++ code. Since portions of empirical are designed to work outside of emscripten we can test those bits without too much of an issue.

I don't know of a good way to metric the code coverage of the emscripten-specific components, though.

**Comment by emilydolson** [comment]
I did a quick google search for "gcov emscripten.' It didn't given me anything particularly informative, but it did turn up this file in the emscripten repo: https://github.com/kripken/emscripten-fastcomp/blob/master/lib/IR/GCOV.cpp. Which kind of implies that gcov will work with emscripten too (and maybe anything based on LLVM?). 

I'm not sure how well that will measure our coverage of javascript embedded in C++ code, though.

**Comment by bocajnotnef** [comment]
Interesting. The intro page I read (
https://gcc.gnu.org/onlinedocs/gcc/Gcov-Intro.html#Gcov-Intro) stated that
gcov will only play nice with gcc-compiled materials--if it'll work on
LLVM-based compilers then that would be fantastic.

I'll do some more research.

**Comment by mercere99** [comment]
I've been surprised and impressed with the number of packages I've tried with Emscripten and they just worked.  That said, it would be surprising that a package that did not work with Clang did work with Emscripten.

As much as possible, I try and make Empirical components usable outside of Emscripten.  Only those that are built to directly interface with JavaScript (most of which are UI-related) won't.

**Comment by mercere99** [comment]
I wonder if we could also use a web testing framework to cover the Emscripten-only portions of Empirical.  It won't be nearly as good as directly unit tests, but much better than none at all.

**Comment by bocajnotnef** [comment]
It's certainly worth a shot. Gcov is (evidently) rather popular so if it
just works then that'd be a big plus.

**Comment by bocajnotnef** [comment]
@mercere99 I assume that we're looking to migrate everything out of the UTests framework within Empirical to another manager, i.e. Catch, and not just add additional tests using that external framework, yes?

**Comment by mercere99** [comment]
Yep -- The UTest directory was a quick and dirty way of making sure that I had SOME tests in place.

**Comment by bocajnotnef** [comment]
Excellent, cool. Just making sure I had the right idea.

**Comment by bocajnotnef** [comment]
While migrating macro tests:

```
testing_tests.cpp:1312: FAILED:
  REQUIRE( result == "(a , b , c , d)" )
with expansion:
  ", (a , b , c , d)" == "(a , b , c , d)"
```

from line:
```
EMP_TEST_MACRO( EMP_ARGS_TO_PACKS_1(4, a,b,c,d,e,f,g), "(a , b , c , d)" );
```

**Comment by bocajnotnef** [comment]
- [ ] Investigate Sauce Labs' UI testing for web interfaces

**Comment by bocajnotnef** [comment]
See: https://saucelabs.com/opensauce/

**Comment by bocajnotnef** [comment]
implementing saucelabs has been migrated to #24, otherwise this is resolved by #12 

Closing #1
